File: software/src/ros4crs/tools/opt_sys_id/src/opt_sys_id/sensor_model/wheel_encoder_sensor_model.py
import copy

import numpy as np
import casadi as ca

from .sensor_model import SensorModel


class WheelEncoderSensorModel(SensorModel):
    """Sensor model for wheel encoders."""

    num_measurements = 4

    def __init__(self, wheel_radius_param, lr_param, lf_param, width_param):
        """
        Initialize the wheel encoder sensor model.

        Args:
            wheel_radius_param: Key of the parameter for the wheel radius
            lr_param: Key of the parameter for lr
            lf_param: Key of the parameter for lf
            width_param: Key of the parameter for the width
        """
        self.wheel_radius_param = wheel_radius_param
        self.lr_param = lr_param
        self.lf_param = lf_param
        self.width_param = width_param

    def calc_y(
        self,
        x: np.array,
        u: np.array,
        numerical: bool = False,
        params: np.ndarray = None,
    ):
        vx = x[3]
        vy = x[4]
        omega = x[5]

        delta = u[1]
        p_width = params[self.width_param]
        p_lr = params[self.lr_param]
        p_lf = params[self.lf_param]
        p_wr = params[self.wheel_radius_param]

        # speed of each wheel for wheel encoder model
        vx_rr = vx + 0.5 * p_width * omega
        vx_rl = vx - 0.5 * p_width * omega

        vx_fr = vx + 0.5 * p_width * omega
        vy_fr = vy + p_lf * omega

        vx_fl = vx - 0.5 * p_width * omega
        vy_fl = vy + p_lf * omega

        vx_wheel_fr = vx_fr * ca.cos(delta) + vy_fr * ca.sin(delta)
        vx_wheel_fl = vx_fl * ca.cos(delta) + vy_fl * ca.sin(delta)
        vx_wheel_rl = vx_rl
        vx_wheel_rr = vx_rr

        if numerical:
            return np.array([vx_wheel_fl, vx_wheel_fr, vx_wheel_rl, vx_wheel_rr]) / p_wr
        else:
            return ca.vertcat(vx_wheel_fl, vx_wheel_fr, vx_wheel_rl, vx_wheel_rr) / p_wr

File: software/src/crs/sensor_models/car/pacejka_sensor_model/src/wheel_encoder_sensor_model.cpp

#include "pacejka_sensor_model/wheel_encoder_sensor_model.h"
#include <dynamic_models/utils/data_conversion.h>

namespace crs_sensor_models
{
namespace pacejka_sensor_models
{
WheelEncoderSensorModel::WheelEncoderSensorModel(double wheel_radius, double lf, double car_width,
                                                 const Eigen::Matrix4d& R)
  : SensorModel(4, WheelEncoderSensorModel::SENSOR_KEY)  // Measurement dimension is three
{
  assert(wheel_radius > 0 &&
         "Invalid Wheel radius for WheelEncoderSensorModel. Make sure it is set in the model parameter config!");

  assert(car_width > 0 &&
         "Invalid car_width for WheelEncoderSensorModel. Make sure it is set in the model parameter config!");

  using namespace casadi;

  // The speed in each wheel frame is calculated as
  // v_wheel = v_car + omega x r_b_wheel #(r_b_wheel is the position of the wheel in the body frame)
  // The back wheels align with x axis of the body frame so the wheel speed in wheel turn direction
  // is the same as the car speed in x direction.
  // The front wheels are turned by steer angle so the wheel speed in wheel turn direction is
  // cos(steer) * v_x_wheel + sin(steer) * v_y_wheel

  auto vx = state_mx[3];
  auto vy = state_mx[4];
  auto dyaw = state_mx[5];
  auto steer = input_mx[1];

  auto vx_fr = vx + 0.5 * car_width * dyaw;
  auto vy_fr = vy + lf * dyaw;
  auto vx_wheel_fr = vx_fr * cos(steer) + vy_fr * sin(steer);

  auto vx_fl = vx - 0.5 * car_width * dyaw;
  auto vy_fl = vy + lf * dyaw;
  auto vx_wheel_fl = vx_fl * cos(steer) + vy_fl * sin(steer);

  auto vx_wheel_rl = vx - 0.5 * car_width * dyaw;
  auto vx_wheel_rr = vx + 0.5 * car_width * dyaw;

  // Unit of wheel speed is rad/s
  // front left, front right, rear left, rear right
  std::vector<casadi::MX> measured_states_mx = { vx_wheel_fl / wheel_radius, vx_wheel_fr / wheel_radius,
                                                 vx_wheel_rl / wheel_radius, vx_wheel_rr / wheel_radius };

  measurement_function = casadi::Function("applyMeasurementModel", state_and_input_mx, measured_states_mx);

  // Define jacobian function using casadi
  // This sets the jacobian_fn directly from the measurement_function
  setJacobianFromMeasFnc(measurement_function);

  setR(R);
}

const std::string WheelEncoderSensorModel::SENSOR_KEY = "wheel_encoders";

}  // namespace pacejka_sensor_models
}  // namespace crs_sensor_models
